hw3/hw3_group2.hs

```diff
@@ -1,3 +1,39 @@
 module Hw3_group2 where
 
--- TODO
\ No newline at end of file
+-- 3 Stack language
+
+-- abstract syntax and supported operations
+type S = [Op]
+data Op = LD Int | ADD | SWAP | DUP
+
+-- stack is a list of integers
+type Stack = [Int]
+
+-- result of operation can be either a stack or a failure string
+data Result a = Success a | Fail String
+semOp :: Op -> Stack -> Result Stack
+
+-- load given int to top of stack
+semOp (LD x) stack = Success (x : stack)
+
+-- add the first 2 elements on the stack or error if fewer than 2 elements present
+semOp ADD [] = Fail ("Error: empty stack, cannot add") 
```

Comment:
You don't strictly need the empty list case for this or SWAP - the _ case will catch anything that doesn't match the two-element case, including empty.

Reply:
Good catch, removed both.